plugins/audio-s5gen2/fu-audio-s5gen2-firmware.c

```diff
@@ -0,0 +1,158 @@
+	FuFirmware parent_instance;
+	guint32 file_id; /* generated ID unique for the firmware */
+	guint8 protocol_ver;
+	gchar *device_variant;
+};
+
+G_DEFINE_TYPE(FuQcS5gen2Firmware, fu_qc_s5gen2_firmware, FU_TYPE_FIRMWARE)
+
+guint8
+fu_qc_s5gen2_firmware_get_protocol_version(FuQcS5gen2Firmware *self)
+{
+	return self->protocol_ver;
+}
+
+guint32
+fu_qc_s5gen2_firmware_get_id(FuQcS5gen2Firmware *self)
+{
+	return self->file_id;
+}
+
+static void
+fu_qc_s5gen2_firmware_export(FuFirmware *firmware, FuFirmwareExportFlags flags, XbBuilderNode *bn)
+{
+	FuQcS5gen2Firmware *self = FU_QC_S5GEN2_FIRMWARE(firmware);
+	fu_xmlb_builder_insert_kv(bn, "device_variant", self->device_variant);
+	fu_xmlb_builder_insert_kx(bn, "protocol_version", self->protocol_ver);
+	fu_xmlb_builder_insert_kx(bn, "generated_file_id", self->file_id);
+}
+
+static gboolean
+fu_qc_s5gen2_firmware_validate(FuFirmware *firmware,
+			       GInputStream *stream,
+			       gsize offset,
+			       GError **error)
+{
+	return fu_struct_qc_fw_update_hdr_validate_stream(stream, offset, error);
+}
+
+static gboolean
+fu_qc_s5gen2_firmware_parse(FuFirmware *firmware,
+			    GInputStream *stream,
+			    gsize offset,
+			    FwupdInstallFlags flags,
+			    GError **error)
+{
+	FuQcS5gen2Firmware *self = FU_QC_S5GEN2_FIRMWARE(firmware);
+	gsize streamsz = 0;
+	const guint8 *device_variant;
+	gsize config_offset = 26;
+	guint16 config_ver;
+
+	/* FIXME: deal with encrypted? */
+	g_autoptr(GByteArray) hdr = NULL;
+	g_autofree gchar *ver_str = NULL;
+
+	if (!fu_input_stream_size(stream, &streamsz, error))
+		return FALSE;
+
+	hdr = fu_struct_qc_fw_update_hdr_parse_stream(stream, offset, error);
+
+	if (hdr == NULL)
+		return FALSE;
+
+	/* protocol version */
+	self->protocol_ver = fu_struct_qc_fw_update_hdr_get_protocol(hdr) - '0';
+
+	device_variant = fu_struct_qc_fw_update_hdr_get_dev_variant(hdr, NULL);
+	self->device_variant = g_strndup((gchar *)device_variant, 8);
+
+	config_offset += fu_struct_qc_fw_update_hdr_get_upgrades(hdr) * 4;
+
+	if (!fu_input_stream_read_u16(stream, config_offset, &config_ver, G_BIG_ENDIAN, error))
+		return FALSE;
+
+	ver_str = g_strdup_printf("%u.%u.%u",
+				  fu_struct_qc_fw_update_hdr_get_major(hdr),
+				  fu_struct_qc_fw_update_hdr_get_minor(hdr),
+				  config_ver);
+
+	fu_firmware_set_version(firmware, ver_str);
+
+	if (!fu_firmware_set_stream(firmware, stream, error))
+		return FALSE;
+
+	if (!fu_input_stream_compute_crc32(stream, &self->file_id, 0xEDB88320, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static GByteArray *
+fu_qc_s5gen2_firmware_write(FuFirmware *firmware, GError **error)
+{
+	FuQcS5gen2Firmware *self = FU_QC_S5GEN2_FIRMWARE(firmware);
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GBytes) fw = NULL;
+
+	/* data first */
+	fw = fu_firmware_get_bytes_with_patches(firmware, error);
+	if (fw == NULL)
+		return NULL;
+
+	/* TODO: write to the buffer with the correct format */
+	g_assert(self != NULL);
```

Comment:
I don't think we want a `g_assert` in the codebase.

Reply:
Good catch! Thank you.